### DISCUSSION TITLE
how to insert an object at a certain place

### DISCUSSION THREAD

**Discussion by movalz** [schibsted/jslt#277] [Q&A] [answered]
I have following input file:

```
{
  "items": [
	{
		"name": "item1",
		"labelKey": "label.item1"
	},
	{
		"name": "item2",
		"labelKey": "label.item2"
	},
	{
		"name": "item3",
		"labelKey": "label.item3"
	}
	]
}
```

and want to insert an additional item between "item2" and "item3":

```
{
  "items": [
	{
		"name": "item1",
		"labelKey": "label.item1"
	},
	{
		"name": "item2",
		"labelKey": "label.item2"
	},
	{
		"name": "item2.1",
		"labelKey": "label.item2.1"
	},
	{
		"name": "item3",
		"labelKey": "label.item3"
	}
	]
}
```

How can I do this? I have tried following transform.jslt file:

```
{
  "items": [for (.items)
  if (.name == "item2")
  {
    "name": "item2.1",
    "labelKey": "label.item2.1"
  }
  else
  {
    *: .
  }
  ]
}
```

but it replaces "item2" by "item2.1" which is understandable because of the "else" statement. What I want to do is either to output "item2" inside the if block (additionally to "item2.1") or an "endif" instead of "else" so that "item2" is printed when the if clause is true (then the if clause has to be changed to if(.name == "item1")).

Any suggestions how to handle this?

**Comment by larsga** [comment]
This snipped will solve the exact example you're describing, although I guess it may not be general enough for your real use case. But it might be enough to show you how to solve it.

```
{"items" : .items[ : 2] + [{
    "name": "item2.1",
    "labelKey": "label.item2.1"
  }] + .items[2 : ]
}
```

**Reply by movalz** [reply]
Thank you for the response. This functions in the concrete case, but not if "item2" isn't the second item:
```
{
  "items": [
	{
		"name": "item2",
		"labelKey": "label.item2"
	},
	{
		"name": "item1",
		"labelKey": "label.item1"
	},
	{
		"name": "item3",
		"labelKey": "label.item3"
	}
	]
}
```
I.e. I want to insert the new item behind "item2" and not "behind position 2":
```
{
  "items": [
	{
		"name": "item2",
		"labelKey": "label.item2"
	},
	{
		"name": "item2.1",
		"labelKey": "label.item2.1"
	},
	{
		"name": "item1",
		"labelKey": "label.item1"
	},
	{
		"name": "item3",
		"labelKey": "label.item3"
	}
	]
}
```
So I need an "if" expression which checks .name == "item2" for determining the position to add "item2.1".

**Comment by larsga** [comment, marked as answer]
The trick is to search for the index of that item, then use that in the expression above to cut the array at the right point. 

This should work:

```
[for (zip-with-index(.items))
  .index if (.value.name == "item2")][0]
```

**Comment by movalz** [comment]
Thank you for the tip. I have solved it. JSLT code:

```
def indexOfItemWithName(iname)
  [for (zip-with-index(.items))
    .index if (.value.name == $iname)]

let idxItem2 = indexOfItemWithName("item2")[0]
let adjItems = .items[ : $idxItem2 + 1] + [{
    "name": "item2.1",
    "labelKey": "label.item2.1"
  }] + .items[$idxItem2 + 1 : ]

{
  "items" : [for ($adjItems)
    {
      *: .
    }
  ]
}
```
